File: ats/ax/classification/temporal.go
package classification

import (
	"sort"
	"time"
)

// TemporalAnalyzer analyzes temporal patterns in claims
type TemporalAnalyzer struct {
	config TemporalConfig
}

// NewTemporalAnalyzer creates a new temporal analyzer with configuration
func NewTemporalAnalyzer(config TemporalConfig) *TemporalAnalyzer {
	return &TemporalAnalyzer{
		config: config,
	}
}

// TemporalPattern represents different temporal relationships between claims
type TemporalPattern string

const (
	TemporalSimultaneous TemporalPattern = "simultaneous" // Within verification window
	TemporalSequential   TemporalPattern = "sequential"   // Clear time ordering
	TemporalOverlapping  TemporalPattern = "overlapping"  // Some temporal overlap
	TemporalDistributed  TemporalPattern = "distributed"  // Spread over long period
)

// ClaimTiming represents timing information for a claim
type ClaimTiming struct {
	Actor     string
	Timestamp time.Time
	Predicate string
}

// AnalyzeTemporalPattern determines the temporal pattern of claims
func (ta *TemporalAnalyzer) AnalyzeTemporalPattern(timings []ClaimTiming) TemporalPattern {
	if len(timings) <= 1 {
		return TemporalSimultaneous
	}

	// Sort by timestamp
	sort.Slice(timings, func(i, j int) bool {
		return timings[i].Timestamp.Before(timings[j].Timestamp)
	})

	earliest := timings[0].Timestamp
	latest := timings[len(timings)-1].Timestamp
	totalSpan := latest.Sub(earliest)

	// Check if all claims are within verification window
	if totalSpan <= ta.config.VerificationWindow {
		return TemporalSimultaneous
	}

	// Check if claims are clearly sequential with gaps
	if ta.hasSequentialGaps(timings) {
		return TemporalSequential
	}

	// Check if spread over long period
	if totalSpan > ta.config.EvolutionWindow {
		return TemporalDistributed
	}

	return TemporalOverlapping
}

// hasSequentialGaps checks if there are clear gaps between claim groups
func (ta *TemporalAnalyzer) hasSequentialGaps(timings []ClaimTiming) bool {
	if len(timings) < 2 {
		return false
	}

	// Look for gaps larger than verification window
	for i := 1; i < len(timings); i++ {
		gap := timings[i].Timestamp.Sub(timings[i-1].Timestamp)
		if gap > ta.config.VerificationWindow {
			return true
		}
	}

	return false
}

// IsSimultaneous checks if two timestamps are within verification window
func (ta *TemporalAnalyzer) IsSimultaneous(t1, t2 time.Time) bool {
	diff := t1.Sub(t2)
	if diff < 0 {
		diff = -diff
	}
	return diff <= ta.config.VerificationWindow
}

// IsEvolutionTimespan checks if timespan suggests natural evolution
func (ta *TemporalAnalyzer) IsEvolutionTimespan(t1, t2 time.Time) bool {
	diff := t1.Sub(t2)
	if diff < 0 {
		diff = -diff
	}
	return diff >= ta.config.VerificationWindow && diff <= ta.config.EvolutionWindow
}

// IsObsolete checks if a timestamp is too old to be relevant
func (ta *TemporalAnalyzer) IsObsolete(timestamp time.Time) bool {
	age := time.Since(timestamp)
	return age > ta.config.ObsolescenceWindow
}

// CalculateRecencyScore returns a score (0.0-1.0) based on how recent a timestamp is
func (ta *TemporalAnalyzer) CalculateRecencyScore(timestamp time.Time) float64 {
	age := time.Since(timestamp)

	// Recent claims (within verification window) get full score
	if age <= ta.config.VerificationWindow {
		return 1.0
	}

	// Claims within evolution window get declining score
	if age <= ta.config.EvolutionWindow {
		ratio := float64(age) / float64(ta.config.EvolutionWindow)
		return 1.0 - (ratio * 0.5) // Decline from 1.0 to 0.5
	}

	// Very old claims get low score but not zero
	if age <= ta.config.ObsolescenceWindow {
		ratio := float64(age) / float64(ta.config.ObsolescenceWindow)
		return 0.5 - (ratio * 0.4) // Decline from 0.5 to 0.1
	}

	// Obsolete claims get minimal score
	return 0.1
}

// GroupByTimeWindow groups claims into temporal windows
func (ta *TemporalAnalyzer) GroupByTimeWindow(timings []ClaimTiming) [][]ClaimTiming {
	if len(timings) == 0 {
		return nil
	}

	// Sort by timestamp
	sort.Slice(timings, func(i, j int) bool {
		return timings[i].Timestamp.Before(timings[j].Timestamp)
	})

	var groups [][]ClaimTiming
	currentGroup := []ClaimTiming{timings[0]}

	for i := 1; i < len(timings); i++ {
		// If within verification window of last item in current group, add to group
		lastInGroup := currentGroup[len(currentGroup)-1]
		if ta.IsSimultaneous(timings[i].Timestamp, lastInGroup.Timestamp) {
			currentGroup = append(currentGroup, timings[i])
		} else {
			// Start new group
			groups = append(groups, currentGroup)
			currentGroup = []ClaimTiming{timings[i]}
		}
	}

	// Add final group
	groups = append(groups, currentGroup)
	return groups
}

// GetMostRecentClaim returns the claim with the most recent timestamp
func (ta *TemporalAnalyzer) GetMostRecentClaim(timings []ClaimTiming) ClaimTiming {
	if len(timings) == 0 {
		return ClaimTiming{}
	}

	mostRecent := timings[0]
	for _, timing := range timings[1:] {
		if timing.Timestamp.After(mostRecent.Timestamp) {
			mostRecent = timing
		}
	}

	return mostRecent
}

// CalculateTemporalConfidence returns confidence based on temporal patterns
func (ta *TemporalAnalyzer) CalculateTemporalConfidence(timings []ClaimTiming) float64 {
	if len(timings) <= 1 {
		return 1.0
	}

	pattern := ta.AnalyzeTemporalPattern(timings)

	switch pattern {
	case TemporalSimultaneous:
		return 0.9 // High confidence for simultaneous verification
	case TemporalSequential:
		return 0.8 // Good confidence for clear evolution
	case TemporalOverlapping:
		return 0.6 // Medium confidence for overlapping claims
	case TemporalDistributed:
		return 0.4 // Lower confidence for distributed claims
	default:
		return 0.5
	}
}

File: crates/qntx-core/src/classify/temporal.rs
//! Temporal pattern analysis for claim classification
//!
//! Analyzes time relationships between claims to determine patterns
//! like simultaneous verification, sequential evolution, or distributed claims.

use serde::{Deserialize, Serialize};

/// Configurable time windows for temporal classification.
/// All values are in milliseconds.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TemporalConfig {
    /// Window within which claims are considered simultaneous (default: 60_000ms = 1 minute)
    pub verification_window_ms: i64,
    /// Window within which claims suggest natural evolution (default: 86_400_000ms = 24 hours)
    pub evolution_window_ms: i64,
    /// Window beyond which claims are considered obsolete (default: 31_536_000_000ms = 1 year)
    pub obsolescence_window_ms: i64,
}

impl Default for TemporalConfig {
    fn default() -> Self {
        Self {
            verification_window_ms: 60_000,         // 1 minute
            evolution_window_ms: 86_400_000,        // 24 hours
            obsolescence_window_ms: 31_536_000_000, // ~365 days
        }
    }
}

/// Temporal pattern between a set of claims
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum TemporalPattern {
    /// All claims within verification window
    Simultaneous,
    /// Clear time ordering with gaps
    Sequential,
    /// Some temporal overlap
    Overlapping,
    /// Spread over a long period
    Distributed,
}

impl std::fmt::Display for TemporalPattern {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Simultaneous => write!(f, "simultaneous"),
            Self::Sequential => write!(f, "sequential"),
            Self::Overlapping => write!(f, "overlapping"),
            Self::Distributed => write!(f, "distributed"),
        }
    }
}

/// Timing information for a single claim
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClaimTiming {
    pub actor: String,
    pub timestamp_ms: i64,
    pub predicate: String,
}

/// Temporal analyzer for claim classification
pub struct TemporalAnalyzer {
    config: TemporalConfig,
}

impl TemporalAnalyzer {
    pub fn new(config: TemporalConfig) -> Self {
        Self { config }
    }

    /// Determine the temporal pattern of a set of claims
    pub fn analyze_pattern(&self, timings: &[ClaimTiming]) -> TemporalPattern {
        if timings.len() <= 1 {
            return TemporalPattern::Simultaneous;
        }

        let mut timestamps: Vec<i64> = timings.iter().map(|t| t.timestamp_ms).collect();
        timestamps.sort();

        let earliest = timestamps[0];
        let latest = timestamps[timestamps.len() - 1];
        let total_span = latest - earliest;

        if total_span <= self.config.verification_window_ms {
            return TemporalPattern::Simultaneous;
        }

        if self.has_sequential_gaps(&timestamps) {
            return TemporalPattern::Sequential;
        }

        if total_span > self.config.evolution_window_ms {
            return TemporalPattern::Distributed;
        }

        TemporalPattern::Overlapping
    }

    /// Check if timestamps have clear gaps (> verification window) between consecutive values
    fn has_sequential_gaps(&self, sorted_timestamps: &[i64]) -> bool {
        for i in 1..sorted_timestamps.len() {
            let gap = sorted_timestamps[i] - sorted_timestamps[i - 1];
            if gap > self.config.verification_window_ms {
                return true;
            }
        }
        false
    }

    /// Check if two timestamps are within the verification window
    pub fn is_simultaneous(&self, t1: i64, t2: i64) -> bool {
        (t1 - t2).unsigned_abs() <= self.config.verification_window_ms as u64
    }

    /// Check if a timespan suggests natural evolution
    pub fn is_evolution_timespan(&self, t1: i64, t2: i64) -> bool {
        let diff = (t1 - t2).unsigned_abs() as i64;
        diff >= self.config.verification_window_ms && diff <= self.config.evolution_window_ms
    }

    /// Calculate recency score (0.0-1.0) based on age relative to a reference time.
    /// `now_ms` is the current time in milliseconds.
    pub fn recency_score(&self, timestamp_ms: i64, now_ms: i64) -> f64 {
        let age = now_ms - timestamp_ms;
        if age < 0 {
            return 1.0; // future timestamps get full score
        }

        if age <= self.config.verification_window_ms {
            return 1.0;
        }

        if age <= self.config.evolution_window_ms {
            let ratio = age as f64 / self.config.evolution_window_ms as f64;
            return 1.0 - (ratio * 0.5); // 1.0 -> 0.5
        }

        if age <= self.config.obsolescence_window_ms {
            let ratio = age as f64 / self.config.obsolescence_window_ms as f64;
            return 0.5 - (ratio * 0.4); // 0.5 -> 0.1
        }

        0.1 // obsolete
    }

    /// Calculate temporal confidence (0.0-1.0) based on the temporal pattern
    pub fn temporal_confidence(&self, timings: &[ClaimTiming]) -> f64 {
        if timings.len() <= 1 {
            return 1.0;
        }

        match self.analyze_pattern(timings) {
            TemporalPattern::Simultaneous => 0.9,
            TemporalPattern::Sequential => 0.8,
            TemporalPattern::Overlapping => 0.6,
            TemporalPattern::Distributed => 0.4,
        }
    }

    /// Get the most recent timestamp from a set of timings
    pub fn most_recent(&self, timings: &[ClaimTiming]) -> Option<i64> {
        timings.iter().map(|t| t.timestamp_ms).max()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn default_analyzer() -> TemporalAnalyzer {
        TemporalAnalyzer::new(TemporalConfig::default())
    }

    fn timing(actor: &str, ts: i64) -> ClaimTiming {
        ClaimTiming {
            actor: actor.to_string(),
            timestamp_ms: ts,
            predicate: "test".to_string(),
        }
    }

    #[test]
    fn single_claim_is_simultaneous() {
        let ta = default_analyzer();
        let timings = vec![timing("alice", 1000)];
        assert_eq!(ta.analyze_pattern(&timings), TemporalPattern::Simultaneous);
    }

    #[test]
    fn claims_within_verification_window_are_simultaneous() {
        let ta = default_analyzer();
        let timings = vec![timing("alice", 1000), timing("bob", 50_000)]; // 49s apart
        assert_eq!(ta.analyze_pattern(&timings), TemporalPattern::Simultaneous);
    }

    #[test]
    fn claims_with_gaps_are_sequential() {
        let ta = default_analyzer();
        let timings = vec![
            timing("alice", 0),
            timing("alice", 120_000), // 2 min gap
        ];
        assert_eq!(ta.analyze_pattern(&timings), TemporalPattern::Sequential);
    }

    #[test]
    fn claims_with_large_gap_are_sequential() {
        // With 2 claims and a gap > verification window, sequential wins
        // (sequential is checked before distributed)
        let ta = default_analyzer();
        let timings = vec![
            timing("alice", 0),
            timing("bob", 100_000_000), // ~1.15 days gap
        ];
        assert_eq!(ta.analyze_pattern(&timings), TemporalPattern::Sequential);
    }

    #[test]
    fn overlapping_chain_spanning_days_is_distributed() {
        // Many claims within verification window of neighbors but spanning > evolution window total.
        // Each consecutive pair is 50s apart (within 60s verification window), but total span = 250s * 346 > 24h.
        // Simpler: use a custom config with a small evolution window for testability.
        let ta = TemporalAnalyzer::new(TemporalConfig {
            verification_window_ms: 60_000,
            evolution_window_ms: 200_000, // 200 seconds
            obsolescence_window_ms: 31_536_000_000,
        });
        // 6 claims, each 50s apart. Total span = 250s > evolution_window (200s).
        // No gap exceeds verification_window (60s), so sequential doesn't match.
        let timings = vec![
            timing("a", 0),
            timing("b", 50_000),
            timing("c", 100_000),
            timing("d", 150_000),
            timing("e", 200_000),
            timing("f", 250_000),
        ];
        assert_eq!(ta.analyze_pattern(&timings), TemporalPattern::Distributed);
    }

    #[test]
    fn recency_score_recent() {
        let ta = default_analyzer();
        let now = 1_000_000;
        // 30 seconds ago
        assert_eq!(ta.recency_score(now - 30_000, now), 1.0);
    }

    #[test]
    fn recency_score_evolution_range() {
        let ta = default_analyzer();
        let now = 1_000_000_000;
        // 12 hours ago (halfway through evolution window)
        let score = ta.recency_score(now - 43_200_000, now);
        assert!(score > 0.5 && score < 1.0);
    }

    #[test]
    fn recency_score_obsolete() {
        let ta = default_analyzer();
        let now = 100_000_000_000;
        // 2 years ago
        let score = ta.recency_score(now - 63_072_000_000, now);
        assert_eq!(score, 0.1);
    }

    #[test]
    fn temporal_confidence_simultaneous_high() {
        let ta = default_analyzer();
        let timings = vec![timing("alice", 1000), timing("bob", 2000)];
        assert_eq!(ta.temporal_confidence(&timings), 0.9);
    }
}
